src/rpc_client/mod.rs

```diff
@@ -34,34 +38,129 @@ pub use self::{
+    /// Convert an [`ApiInfo`] to a string
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if let Some(token) = &self.token {
+            token.fmt(f)?;
+            write!(f, ":")?;
+        }
+        self.multiaddr.fmt(f)?;
+        Ok(())
+    }
+}
+
+impl FromStr for ApiInfo {
+    type Err = multiaddr::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(match s.split_once(':') {
+            // token:host
+            Some((jwt, host)) => ApiInfo {
+                multiaddr: host.parse()?,
+                token: Some(jwt.to_owned()),
+            },
+            // host
+            None => ApiInfo {
+                multiaddr: s.parse()?,
+                token: None,
+            },
+        })
+    }
+}
+
+impl ApiInfo {
+    // Update API handle with new (optional) token
+    pub fn set_token(self, token: Option<String>) -> Self {
+        ApiInfo {
+            token: token.or(self.token),
+            ..self
+        }
+    }
+
     // Get API_INFO environment variable if exists, otherwise, use default
-    // multiaddress
-    let api_info = env::var(API_INFO_KEY).unwrap_or_else(|_| DEFAULT_MULTIADDRESS.to_owned());
-
-    let (multiaddr, token) = match api_info.split_once(':') {
-        // Typically this is when a JWT was provided
-        Some((jwt, host)) => (
-            host.parse().expect("Parse multiaddress"),
-            Some(jwt.to_owned()),
-        ),
-        // Use entire API_INFO env var as host string
-        None => (api_info.parse().expect("Parse multiaddress"), None),
-    };
+    // multiaddress. Fails if the environment variable is malformed.
+    pub fn from_env() -> Result<Self, multiaddr::Error> {
+        let api_info = env::var(API_INFO_KEY).unwrap_or_else(|_| DEFAULT_MULTIADDRESS.to_owned());
+        ApiInfo::from_str(&api_info)
+    }
+
+    pub async fn call<T: HasLotusJson>(&self, req: RpcRequest<T>) -> Result<T, JsonRpcError> {
+        let rpc_req = RequestObject::request()
+            .with_method(req.method_name)
+            .with_params(req.params)
+            .with_id(0)
+            .finish();
+
+        let api_url = multiaddress_to_url(&self.multiaddr);
+
+        debug!("Using JSON-RPC v2 HTTP URL: {}", api_url);
 
-    ApiInfo { multiaddr, token }
-});
+        let request = global_http_client().post(api_url).json(&rpc_req);
+        let request = match self.token.as_ref() {
+            Some(token) => request.header(http::header::AUTHORIZATION, token),
+            _ => request,
+        };
+
+        let rpc_res: JsonRpcResponse<T::LotusJson> =
+            request.send().await?.error_for_status()?.json().await?;
+
+        match rpc_res {
+            JsonRpcResponse::Result { result, .. } => Ok(HasLotusJson::from_lotus_json(result)),
+            JsonRpcResponse::Error { error, .. } => Err(error),
+        }
+    }
+}
 
 /// Error object in a response
-#[derive(Deserialize)]
+#[derive(Debug, Deserialize)]
 pub struct JsonRpcError {
```

Comment:
```rust
/// We don't use [`jsonrpc::Error`] because...
```

Reply:
... because we're using a completely different json-rpc library? Arguing the differences between json-rpc crates is out of scope of this PR.